chain/messagepool/messagepool.go

```diff
@@ -1184,11 +1203,6 @@ func (mp *MessagePool) allPending(ctx context.Context) ([]*types.SignedMessage,
 }
 
 func (mp *MessagePool) PendingFor(ctx context.Context, a address.Address) ([]*types.SignedMessage, *types.TipSet) {
-	mp.curTsLk.Lock()
-	defer mp.curTsLk.Unlock()
```

Comment:
This is the key thing to determine. When I read the code, it appears that PendingFor only accesses the variables pending and keyCache. Good point about msgSet potentially being modified on us. What was the expectation for curTsLk ? What was it guarding against, or was it a generic synchronization lock around the message pool?

Reply:
`curTsLk` is making it safe to return the mp.curTs, and afaiu we MUST call mp.pendingFor(ctx, a) while holding that lock, which is needed to make `PendingFor` have some properties:
* The returned messages definitely haven't been included in the returned tipset
* If we have just had a revert, messages from reverted blocks were re-added to the pending set, so that message nonces keep being continuous when you look at the chain+pending messages

There is maybe a way to swap it to an RW lock, but I don't think there's a safe way to get rid of that lock here